### ISSUE TITLE
Mean absolute error?

### ISSUE THREAD

**Issue by erikbern** [dmlc/xgboost#603]
I noticed there's an objective for root mean square error but not for mean absolute error? I'm happy to put together a PR for this if you think this would be useful.

Actually what I think would be even more useful is to generalize this to quantile loss – happy to include that too. My use case is I want to predict a 90% interval. This could be done by training two separate predictors, quantile loss for 0.05 and 0.95.

Let me know if you think this is a bad idea and I won't attempt it :)

**Comment by linuxthink** [comment]
This will definitely be useful. In [scikit-learn's implementation] (http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.GradientBoostingRegressor.html), it has squared, absolute, huber and quantile loss for regression tasks. It would be nice if xgboost also support them.

**Comment by erikbern** [comment]
I started adding this and it seems pretty straightforward. There is one problem though. xgboost computes the second derivative and for quantile loss it's a Dirac delta, i.e. zero everywhere except an infinite spike around x=0.

I'm not sure how the second derivative (Hessian) is used – it doesn't seem to be used in sklearn's implementation

**Comment by erikbern** [comment]
Closing this issue for now – will try to put together a pull request with the implementation. But would love to hear if you have any thoughts about what to do with the hessian

**Comment by erikbern** [comment]
I looked at this some more and I have implemented something.

The problem is again with the Hessian. From what I understand, xgboost uses Newton's method to find the optimum and that's why the Hessian is used.

I see several options
* Instead of mean absolute error, use Huber loss
* Instead of mean absolute error, use something like `log(exp(-x) + exp(x)` which is similar
* Use mean absolute error and make xgboost work with first order methods

I can put together a PR for Huber loss for now, but I'm not super excited about it since it requires a parameter `delta`. But maybe let's start there?

**Comment by tqchen** [comment]
closing for now, also note #736 refactor gives an plugin system https://github.com/dmlc/xgboost/tree/brick/plugin to add in new loss and metrics

**Comment by sachinruk** [comment]
@erikbern I was looking for this exactly. Is there a Hubert loss equivalent for the tilted absolute loss case (i.e. when you want other quantiles besides the median?).

Even if you did get a second derivative is there a batch processing method in XgBoost, otherwise taking the Hessian might be prohibitive.

**Comment by erikbern** [comment]
I implemented Huber but it has the same problems with convergence since the derivative isn't continuous.

I think the only solution would be to support a first-order method that doesn't require the Hessian

**Comment by khotilov** [comment]
The twice differentiable log-cosh loss seems to be a reasonably good option for the xgboost's 2nd order framework.

**Comment by thvasilo** [comment]
Hello @tqchen,

Are [plugins](https://github.com/dmlc/xgboost/tree/master/plugin) still the recommended way to add custom loss functions?

Are plugins only usable through the C++ frontend?
If yes, I guess I won't be able to use my custom loss in R.

I assume I should be using the example of a [custom objective for R](https://github.com/dmlc/xgboost/blob/master/R-package/demo/custom_objective.R) instead.

**Comment by arbackus** [comment]
The log-cosh objective still suffers from the gradient and hessian for very large off-target predictions being constant, therefore resulting in absence of splits. Here is a blog post on the topic, a possible solution and implementation in Python: http://www.bigdatarepublic.nl/regression-prediction-intervals-with-xgboost/
